src/Interpreters/ConcurrentHashJoin.cpp

```diff
@@ -0,0 +1,323 @@
+    }
+
+    ColumnsWithTypeAndName final_columns;
+    MutableColumns mutable_final_columns;
+    NamesAndTypesList names_and_types = dispatched_blocks[0].getNamesAndTypesList();
+    auto types = names_and_types.getTypes();
+    for (auto & dispatched_block : dispatched_blocks)
+    {
+        for (size_t pos = 0; pos < dispatched_block.columns(); ++pos)
+        {
+            auto & from_column = dispatched_block.getByPosition(pos);
+            if (mutable_final_columns.size() <= pos)
+            {
+                mutable_final_columns.emplace_back(from_column.column->cloneEmpty());
+            }
+            if (!from_column.column->empty())
+            {
+                mutable_final_columns[pos]->insertRangeFrom(*from_column.column, 0, from_column.column->size());
+            }
+        }
+    }
+
+    size_t i = 0;
+    for (auto & name_and_type : names_and_types)
+    {
+        final_columns.emplace_back(ColumnWithTypeAndName(std::move(mutable_final_columns[i]), name_and_type.type, name_and_type.name));
+        i += 1;
+    }
+    block = Block(final_columns);
+}
+
+void ConcurrentHashJoin::checkTypesOfKeys(const Block & block) const
+{
+    hash_joins[0]->data->checkTypesOfKeys(block);
+}
+
+void ConcurrentHashJoin::setTotals(const Block & block)
+{
+    if (block)
+    {
+        std::lock_guard lock(totals_mutex);
+        totals = block;
+    }
+}
+
+const Block & ConcurrentHashJoin::getTotals() const
+{
+    return totals;
+}
+
+size_t ConcurrentHashJoin::getTotalRowCount() const
+{
+    size_t res = 0;
+    for (const auto & hash_join : hash_joins)
+    {
+        res += hash_join->data->getTotalRowCount();
+    }
+    return res;
+}
+
+size_t ConcurrentHashJoin::getTotalByteCount() const
+{
+    size_t res = 0;
+    for (const auto & hash_join : hash_joins)
+    {
+        res += hash_join->data->getTotalByteCount();
+    }
+    return res;
+}
+
+bool ConcurrentHashJoin::alwaysReturnsEmptySet() const
+{
+    for (const auto & hash_join : hash_joins)
+    {
+        if (!hash_join->data->alwaysReturnsEmptySet() || !hash_join->pending_right_blocks.empty())
+            return false;
+    }
+    return true;
+}
+
+std::shared_ptr<NotJoinedBlocks> ConcurrentHashJoin::getNonJoinedBlocks(
+        const Block & /*left_sample_block*/, const Block & /*result_sample_block*/, UInt64 /*max_block_size*/) const
+{
+    if (table_join->strictness() == ASTTableJoin::Strictness::Asof ||
+        table_join->strictness() == ASTTableJoin::Strictness::Semi ||
+        !isRightOrFull(table_join->kind()))
+    {
+        return {};
+    }
+    throw Exception(ErrorCodes::LOGICAL_ERROR, "Invalid join type. join kind: {}, strictness: {}", table_join->kind(), table_join->strictness());
+}
+
+std::shared_ptr<ExpressionActions> ConcurrentHashJoin::buildHashExpressionAction(const Block & block, const Strings & based_columns_names, Strings & hash_columns_names)
```

Comment:
Why do we need such complex way to calculate hash? We can just access data in Block and calculate what we need

Reply:
Do you have any example to calculate hash from block data directly ?
This is simplest the way that I have found

Reply:
Found that , `Blocks DistributedSink::splitBlock(const Block & block)` use a simillar way by using expression action

Reply:
@lgbo-ustc See changes in https://github.com/ClickHouse/ClickHouse/tree/concurrent_join (Also you may allow to push this changes directly to PR [link](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)). We don't need create actions, just calculate hash. Also cityHash is to heavy, some weak faster hash would be fine.

Reply:
updated